Add a few type summaries for StyleLength types.

sammygill opened the pull request:
#### 7330772d995fe52703882143d0e37717ef941632
<pre>
Add a few type summaries for StyleLength types.
<a href="https://bugs.webkit.org/show_bug.cgi?id=298062">https://bugs.webkit.org/show_bug.cgi?id=298062</a>
<a href="https://rdar.apple.com/159391129">rdar://159391129</a>

Reviewed by Alan Baradlay.

With the introduction of the new strongly typed Style system, it has
become a bit cumbersome to figure out what the underlying length is
while you are in the middle of a debug session. For example, if you want
to find out what the type is, you would need to query via the isFixed(),
isPercent(), etc. functions until you find the right type.

This patch helps improve the debugging experience a bit by adding some
type summaries to a few of these new length types (PreferredSize,
MinimumSize, and MaximumSize) to provide a summary of the length in a
nice and readable format. I am only adding summaries for these types
specifically in this initial patch since they are fairly common but it
should hopefully be straightforward to build upon this and add summaries
for the rest as needed. One thing to note is that the calc summary
just returns &quot;calc,&quot; but this is the same as the old Length summary
provider.

The structure of the summary providers is based upon the overall
architecture of LengthWrapperBase and its derived classes.
Implementations of LengthWrapperBase provide any keyword values via
template parameters, which end up influencing how its opaque type is
mapped. For example, the first keyword of the length is mapped to 1, the
second is mapped to 2, etc. Fixed is mapped to value right after
the last keyword, percentage is mapped to right after fixed, and calc
is mapped to fixed + 2.

This means that we can abstract out most of the logic for these type
summaries since the only thing that makes the behavior of the
implementations is which keywords we support.

* Tools/lldb/lldb_webkit.py:
(WebCoreStyleLengthProviderBase):
This base class is used to contain most of the logic/API that is
required to analyze the actual type. All the derived class needs to do
is override the list of keywords for whatever is supported.

(WebCoreStyleLengthProviderBase.__init__):
(WebCoreStyleLengthProviderBase.has_quirk):
(WebCoreStyleLengthProviderBase.has_keywords):
(WebCoreStyleLengthProviderBase.fixed_type_value):
(WebCoreStyleLengthProviderBase.percentage_type_value):
(WebCoreStyleLengthProviderBase.calculated_type_value):
Maps the same logic as in LengthWrapperBase for the mapping of these
values in relation to the keywords.

(addSummaryProviderForWebCoreStyleLength):
This helper function is used to add the type summaries for
implementation LengthWrapperBase. All that is needed is the type that a
summary is needed for (e.g. WebCore::Style::PreferredSize) along with a
list of any keywords that it supports.

Canonical link: <a href="https://commits.webkit.org/299299@main">https://commits.webkit.org/299299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fdb799c25f0ff7eb5ec7c0d682a61ac21b2f389a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118529 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38210 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28861 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124701 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70588 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b355ea29-239c-4d07-8bc8-e162ab30b78c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120407 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46792 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89965 "Found 1 new test failure: http/wpt/service-workers/service-worker-spinning-activate.https.html (failure)") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59530 "Build is in progress. Recent messages:") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5b606e12-90fe-4f15-ba77-eeb6c331241a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121482 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106271 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70469 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cc96e06e-90fb-4cc0-afa9-168f14c3efc3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24382 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68360 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127767 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45436 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34277 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98609 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45800 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102492 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98393 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25014 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43821 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21812 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45306 "Built successfully") | | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44769 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48116 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46456 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Tools/lldb/lldb_webkit.py:
```diff
@@ -113,6 +204,39 @@ def __lldb_init_module(debugger, dict):
+        "WebKitFillAvailable",
+        "Intrinsic",
+        "MinIntrinsic"
+    ]
```
will these lists ever differ? If not, can we have a single list?

vitorroriz commented on Tools/lldb/lldb_webkit.py:
```diff
@@ -113,6 +204,39 @@ def __lldb_init_module(debugger, dict):
+        "FitContent",
+        "WebKitFillAvailable",
+        "Intrinsic",
+        "MinIntrinsic"
```
For consistency: other lists have trailing comma (last item)

sammygill committed:
Add a few type summaries for StyleLength types.
https://bugs.webkit.org/show_bug.cgi?id=298062
rdar://159391129

Reviewed by Alan Baradlay.

With the introduction of the new strongly typed Style system, it has
become a bit cumbersome to figure out what the underlying length is
while you are in the middle of a debug session. For example, if you want
to find out what the type is, you would need to query via the isFixed(),
isPercent(), etc. functions until you find the right type.

This patch helps improve the debugging experience a bit by adding some
type summaries to a few of these new length types (PreferredSize,
MinimumSize, and MaximumSize) to provide a summary of the length in a
nice and readable format. I am only adding summaries for these types
specifically in this initial patch since they are fairly common but it
should hopefully be straightforward to build upon this and add summaries
for the rest as needed. One thing to note is that the calc summary
just returns "calc," but this is the same as the old Length summary
provider.

The structure of the summary providers is based upon the overall
architecture of LengthWrapperBase and its derived classes.
Implementations of LengthWrapperBase provide any keyword values via
template parameters, which end up influencing how its opaque type is
mapped. For example, the first keyword of the length is mapped to 1, the
second is mapped to 2, etc. Fixed is mapped to value right after
the last keyword, percentage is mapped to right after fixed, and calc
is mapped to fixed + 2.

This means that we can abstract out most of the logic for these type
summaries since the only thing that makes the behavior of the
implementations is which keywords we support.

* Tools/lldb/lldb_webkit.py:
(WebCoreStyleLengthProviderBase):
This base class is used to contain most of the logic/API that is
required to analyze the actual type. All the derived class needs to do
is override the list of keywords for whatever is supported.

(WebCoreStyleLengthProviderBase.__init__):
(WebCoreStyleLengthProviderBase.has_quirk):
(WebCoreStyleLengthProviderBase.has_keywords):
(WebCoreStyleLengthProviderBase.fixed_type_value):
(WebCoreStyleLengthProviderBase.percentage_type_value):
(WebCoreStyleLengthProviderBase.calculated_type_value):
Maps the same logic as in LengthWrapperBase for the mapping of these
values in relation to the keywords.

(addSummaryProviderForWebCoreStyleLength):
This helper function is used to add the type summaries for
implementation LengthWrapperBase. All that is needed is the type that a
summary is needed for (e.g. WebCore::Style::PreferredSize) along with a
list of any keywords that it supports.

Canonical link: https://commits.webkit.org/299299@main
